File: src/test/java/tfol/FreeVarsTest.java
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotEquals;
import static org.junit.Assert.assertNull;
import org.junit.Test;
import org.junit.BeforeClass;
import org.junit.Ignore;

import fortress.tfol.*;
import java.util.List;
import java.util.Set;
import java.util.ArrayList;
import java.io.*;
import fortress.util.Errors;

public class FreeVarsTest {
    
    Sort A = Sort.mkSortConst("A");
    Sort B = Sort.mkSortConst("B");
    Var x = Term.mkVar("x");
    Var y = Term.mkVar("y");
    Var z = Term.mkVar("z");
    Var c = Term.mkVar("c");
    Var p = Term.mkVar("p");
    Var q = Term.mkVar("q");
    
    Term t1 = Term.mkAnd(
        Term.mkApp("R", x),
        Term.mkNot(
            Term.mkImp(
                Term.mkApp("Q", y),
                Term.mkApp("R", x)
            )
        )
    );
    
    @Test
    public void simpleTerm() {
        assertEquals(Set.of(x, y), t1.freeVarConstSymbolsJava());
        assertEquals(Set.of(x, y), t1.freeVarsJava(Signature.empty()));
    }
    
    @Test
    public void quantifiedTerm() {
        Term t2 = Term.mkForall(List.of(x.of(A), y.of(B)), Term.mkImp(t1, z));
        assertEquals(Set.of(z), t2.freeVarConstSymbolsJava());
        assertEquals(Set.of(z), t2.freeVarsJava(Signature.empty()));
    }
    
    @Test
    public void constantsNotFree() {
        Signature sig = Signature.empty()
            .withSort(A)
            .withConstant(c.of(A))
            .withConstant(p.of(Sort.Bool()));
            
        Term t = Term.mkAnd(
            Term.mkEq(c, x),
            Term.mkImp(p, q)
        );
        
        assertEquals(Set.of(x, q), t.freeVarsJava(sig));
    }
}

File: src/test/scala/operations/FreeVarsTest.scala
import org.scalatest._

import fortress.msfol._
import fortress.operations.TermOps._

class FreeVarsTest extends UnitSuite {
    
    val A = Sort.mkSortConst("A")
    val B = Sort.mkSortConst("B")
    val x = Term.mkVar("x")
    val y = Term.mkVar("y")
    val z = Term.mkVar("z")
    val c = Term.mkVar("c")
    val p = Term.mkVar("p")
    val q = Term.mkVar("q")
    
    val t1 = And(
        App("R", x),
        Not(
            Implication(
                App("Q", y),
                App("R", x)
            )
        )
    )
    
    test("simple term") {
        t1.freeVarConstSymbols should be (Set(x, y))
        t1.freeVars(Signature.empty) should be (Set(x, y))
    }
    
    test("quantified term") {
        val t2 = Forall(Seq(x of A, y of B), Implication(t1, z))
        t2.freeVarConstSymbols should be (Set(z))
        t2.freeVars(Signature.empty) should be (Set(z))
    }
    
    test("constants not free") {
        val sig = Signature.empty
            .withSort(A)
            .withConstant(c of A)
            .withConstant(p of (Sort.Bool))
            
        val t = And(
            Eq(c, x),
            Implication(p, q)
        )
        
        t.freeVarConstSymbols should be (Set(c, x, p, q))
        t.freeVars(sig) should be (Set(x, q))
    }
}
